pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gabagecollector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const ResourceResyncTime = 30 * time.Second
+
+type GarbageCollector struct {
+	restMapper         meta.RESTMapper
+	clientPool         dynamic.ClientPool
+	dirty_queue        *workqueue.Type
+	scannerStores      []cache.Store
+	scannerControllers []*framework.Controller
+}
+
+func NewGarbageCollector(clientPool dynamic.ClientPool, resources []unversioned.GroupVersionResource) *GarbageCollector {
+	gc := &GarbageCollector{
+		clientPool:  clientPool,
+		dirty_queue: workqueue.New(),
+		// TODO: should use a dynamic RESTMapper built from the discovery results.
+		restMapper: registered.RESTMapper(),
+	}
+	for _, resource := range resources {
```

Comment:
> We had another primitive idea: we can add an endpoint to the API server, which will take a TypeMeta and/or ObjectMeta as input, and returns the URL for the resource or object. We could make the returned URL point to the canonical version. This feature would be needed by @krousey's work of simplifying kubectl code as well.

Using `TypeMeta` to return a single URL doesn't seem right.  `TypeMeta` uses a `Kind` which can logically map to multiple `Resources`.  At best, such an API returns back an ordered list.

Even if you create such an API, it doesn't given you  a "canonical" resource because you can end up with multiple distinct resources associated with the kind that each need to be garbage collected.

I think we should be very careful about the cardinality assumptions we build in to avoid problems similar to the ones we've had with HPA.